### PR TITLE
Allow only specific spv storage classes for binding decoration

### PR DESCRIPTION
**Pull request by kaizhangNV**
In
https://registry.khronos.org/vulkan/specs/1.3/html/chap37.html#VUID-StandaloneSpirv-DescriptorSet-06491

it states that
If a variable is decorated by DescriptorSet or Binding, the Storage class must be UniformConstant, Uniform and StorageBuffer.

So apply this rule to our emit-spirv logic.

### CODE REVIEW

**Comment by kaizhangNV** [comment]
I will add a unit test for this soon.

File: source/slang/slang-emit-spirv.cpp [hunk 1]
```diff
@@ -1791,6 +1791,26 @@ struct SPIRVEmitContext
...
+        }
+        return storageClass;
+    }
+
+    // https://registry.khronos.org/vulkan/specs/1.3/html/chap37.html#VUID-StandaloneSpirv-DescriptorSet-06491
+    // Only UniformConstant, Uniform or StorageBuffer storage class are allowed to be decorated with descriptor
+    // set or binding.
+    static inline bool isBindingAllowed(SpvStorageClass storageClass)
+    {
+        return  (storageClass == SpvStorageClassUniformConstant) ||
```
Comment (csyonghe) [inline source/slang/slang-emit-spirv.cpp:1809]: Use `switch` to list all cases.

File: source/slang/slang-emit-spirv.cpp [hunk 2]
```diff
@@ -2196,12 +2220,15 @@ struct SPIRVEmitContext
             case LayoutResourceKind::UnorderedAccess:
             case LayoutResourceKind::SamplerState:
             case LayoutResourceKind::DescriptorTableSlot:
-                emitOpDecorateBinding(
-                    getSection(SpvLogicalSectionID::Annotations),
-                    nullptr,
-                    varInst,
-                    SpvLiteralInteger::from32(int32_t(index)));
-                if (!isDescirptorSetDecorated)
+                if (isBindingDecorationAllowed)
```
Comment (csyonghe) [inline source/slang/slang-emit-spirv.cpp:2223]: just do:
```
if (!isBindingDecorationAllowed)
     break;
```

File: source/slang/slang-emit-spirv.cpp (hunk 1 above)
Comment (kaizhangNV) [reply source/slang/slang-emit-spirv.cpp:1809]: You want me to list all storage classes or just these 3? There are lots of cases.

Comment (csyonghe) [reply source/slang/slang-emit-spirv.cpp]: Just turn this into a switch statement and list the three cases. No need to list everything.

Comment (kaizhangNV) [reply source/slang/slang-emit-spirv.cpp]: OK, I see.
But why we prefer the switch..case to this logic expression? Is just a performance concern?

File: source/slang/slang-emit-spirv.cpp (hunk 2 above)
Comment (kaizhangNV) [reply source/slang/slang-emit-spirv.cpp]: act

**Review by csyonghe — APPROVED** [review]

**Pull request merged**